aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,648 @@
+                }
+            }
+        }
+    }
+}
+
+/// Aggregator table - a top-level collection storing aggregators. It holds an
+/// inner table data structure to actually store the aggregators.
+struct AggregatorTable {
+    // All aggregators created with this table.
+    new_aggregators: BTreeSet<u128>,
+    // All aggregators removed from this table.
+    removed_aggregators: BTreeSet<u128>,
+    // Table to store all agregators associated with this `AggregatorTable`.
+    table: BTreeMap<u128, Aggregator>,
+}
+
+impl AggregatorTable {
+    /// Returns a possibly new aggregator instance. If transaction did not
+    /// initialize this aggregator, then the actual value is not known, and the
+    /// aggregator state must be a delta, zero-intialized.
+    fn get_or_create_aggregator(&mut self, key: u128, limit: u128) -> &mut Aggregator {
+        self.table.entry(key).or_insert_with(|| Aggregator {
+            key,
+            state: AggregatorState::PositiveDelta,
+            value: 0,
+            limit,
+        });
+        self.table.get_mut(&key).unwrap()
+    }
+
+    fn num_aggregators(&self) -> u128 {
+        self.table.len() as u128
+    }
+
+    /// Creates and inserts into table a new Aggregator with given `key` and
+    /// `limit`. Since this transaction initializes aggregator, then the actual
+    /// value is known, and the aggregator state must be `Data`. Also,
+    /// aggregators are zero-initialized.
+    fn create_new_aggregator(&mut self, key: u128, limit: u128) {
+        let aggregator = Aggregator {
+            key,
+            state: AggregatorState::Data,
+            value: 0,
+            limit,
+        };
+        self.table.insert(key, aggregator);
+        self.new_aggregators.insert(key);
+    }
+
+    /// Removes an aggregator.
+    fn remove_aggregator(&mut self, key: u128) {
+        // Aggregator no longer in use during this context, so remove it
+        // from the table if it was there.
+        self.table.remove(&key);
+
+        if self.new_aggregators.contains(&key) {
+            // Aggregator has been created by the same transaction. Therefore,
+            // no side-effects.
+            self.new_aggregators.remove(&key);
+        } else {
+            // Otherwise, aggregator has been created somewhere else.
+            self.removed_aggregators.insert(key);
+        }
+    }
+}
+
+/// Top-level struct that is used by the context to maintain the map of all
+/// aggregator tables created throughout the VM session. In theory, only one
+/// is sufficient.
+#[derive(Default)]
+struct AggregatorTableData {
+    aggregator_tables: BTreeMap<TableHandle, AggregatorTable>,
+}
+
+impl AggregatorTableData {
+    /// Returns a mutable reference to the `AggregatorTable` specified by its
+    /// handle. A new aggregator table is created if it doesn't exist in this
+    /// context.
+    fn get_or_create_aggregator_table(
```

Comment:
This is what @gelash mentioned as well: this design implies that we can have multiple tables which allow aggregator creation. Since we limit the table to be stored on the core account only, this can (and will) be dropped.